Source/DafnyCore/DafnyMain.cs

```diff
@@ -64,27 +66,72 @@ public class DafnyFile {
         SourceFileName = filePath;
       } else if (extension == ".dll") {
         IsPrecompiled = true;
-        var asm = Assembly.LoadFile(filePath);
-        string sourceText = null;
-        foreach (var adata in asm.CustomAttributes) {
-          if (adata.Constructor.DeclaringType.Name == "DafnySourceAttribute") {
-            foreach (var args in adata.ConstructorArguments) {
-              if (args.ArgumentType.FullName == "System.String") {
-                sourceText = (string)args.Value;
-              }
-            }
-          }
-        }
 
+        var sourceText = GetDafnySourceAttributeText(filePath);
         if (sourceText == null) { throw new IllegalDafnyFile(); }
         SourceFileName = Path.GetTempFileName();
         File.WriteAllText(SourceFileName, sourceText);
 
       } else {
         throw new IllegalDafnyFile();
       }
+    }
+
+    private static string GetDafnySourceAttributeText(string dllPath) {
+      using var dllFs = new FileStream(dllPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
```

Comment:
Does this result in a reasonable error if `dllPath` is a non-existent file? Or does the earlier code in `DafnyFile` ensure that it exists?

Reply:
It is no worse than the old version that tried to `Assembly.LoadFile` it: you still get a `System.IO.FileNotFoundException`, just with a slightly different stack trace.

It's worth making that cleaner since it's a user error that shouldn't lead to a crash, but it may be non-trivial to fix. Throwing `IllegalDafnyFile` will probably lead to a different less helpful error so it may need some refactoring. I cut #2719 as a follow up.